containers/hawkular-openshift-security-filter/src/main/java/org/hawkular/openshift/auth/org/hawkular/openshift/namespace/NamespaceHandler.java

```diff
@@ -0,0 +1,128 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.openshift.auth.org.hawkular.openshift.namespace;
+
+import static io.undertow.util.StatusCodes.INTERNAL_SERVER_ERROR;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Map;
+
+import org.hawkular.openshift.auth.Utils;
+import org.jboss.logging.Logger;
+
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.HttpString;
+
+/**
+ *
+ */
+public class NamespaceHandler implements HttpHandler {
+
+    private static final Logger log = Logger.getLogger(NamespaceHandler.class);
+
+    private final HttpHandler containerHandler;
+
+    //
+    private static final String HAWKULAR_TENANT_HEADER_NAME = "Hawkular-Tenant";
+    private static final String TOKEN_FILE_LOCATION = "/var/run/secrets/kubernetes.io/serviceaccount/token";
+
+    private static final String KUBERNETES_MASTER_URL_SYSPROP = "KUBERNETES_MASTER_URL";
+    private static final String KUBERNETES_MASTER_URL_DEFAULT = "https://kubernetes.default.svc.cluster.local";
+    static final String KUBERNETES_MASTER_URL =
+            System.getProperty(KUBERNETES_MASTER_URL_SYSPROP, KUBERNETES_MASTER_URL_DEFAULT);
+    //
+
+    private final String token;
+
+    private final NamespaceListener namespaceListener;
+
+    private final Map<String,String> projectOverwrites;
+
+    public NamespaceHandler(HttpHandler containerHandler) {
+        log.info("Creating NamespaceHandler");
+        this.containerHandler = containerHandler;
+
+        try {
+            this.token = getToken();
+            log.info("Received token from " + TOKEN_FILE_LOCATION);
+        } catch (Exception e){
+            log.info("Error retrieving token from file " + TOKEN_FILE_LOCATION);
+            throw new RuntimeException(e);
+        }
+
+        namespaceListener = new NamespaceListener(token);
+        NamespaceOverrideMapper nom = new NamespaceOverrideMapper(token);
+        log.info("Received namespace overrides: " + nom.overrides);
+
+        this.projectOverwrites = nom.overrides;
+    }
+
+    private String getToken() throws IOException {
+        return new String(Files.readAllBytes(Paths.get(TOKEN_FILE_LOCATION)));
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange serverExchange) throws Exception {
+        String originalTenantName = getTenantName(serverExchange);
+        if (originalTenantName == null || originalTenantName.startsWith("_")) {
```

Comment:
Yep, it always has. We use it for our _system tenant, _hawkular_admin tenant, and is used by others for non project based tenants.